### ISSUE TITLE
resource_sql_database_instance fails when maintenance window is not declared.

### ISSUE THREAD

**Issue by gburiola** [terraform-providers/terraform-provider-google#2716]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment
* If an issue is assigned to the "modular-magician" user, it is either in the process of being autogenerated, or is planned to be autogenerated soon. If an issue is assigned to a user, that user is claiming responsibility for the issue. If an issue is assigned to "hashibot", a community member has claimed the issue already.

### Issue summary

Creating a `sql_database_instance` without declaring `settings.maintenance_window` works the first time terraform is run.

Running terraform a second time without changing `.tf` files shows a diff in `settings.maintenance_window`. Applying the changes identified in the second run fail.

### Description

According to the docs, `settings.maintenance_window` subblock should be optional
https://www.terraform.io/docs/providers/google/r/sql_database_instance.html

When you create a new `database_instance` with terraform, it works fine on the first apply, then if you run plan/apply a second time it shows the following diff:

```
~ google_sql_database_instance.NAME
    settings.0.maintenance_window.#: "0" => "1"
```

Trying to apply the diff above fails with:

```
* google_sql_database_instance.NAME: Error, failed to update instance NAME: googleapi: 
  Error 400: This operation isn"t valid for this instance., invalidOperation
```

That is probably because the `PUT` request contains an empty `maintenanceWindow`:

```
plugin.terraform-provider-google_v1.20.0_x4: ---[ REQUEST ]---------------------------------------
plugin.terraform-provider-google_v1.20.0_x4: PUT /sql/v1beta4/projects/PROJECT/instances/NAME?alt=json&prettyPrint=false HTTP/1.1
plugin.terraform-provider-google_v1.20.0_x4: Host: www.googleapis.com
plugin.terraform-provider-google_v1.20.0_x4: User-Agent: google-api-go-client/0.5 Terraform/0.11.7 (+https://www.terraform.io) terraform-provider-google/1.20.0
plugin.terraform-provider-google_v1.20.0_x4: Content-Length: 2773
plugin.terraform-provider-google_v1.20.0_x4: Content-Type: application/json
plugin.terraform-provider-google_v1.20.0_x4: Accept-Encoding: gzip
plugin.terraform-provider-google_v1.20.0_x4:
plugin.terraform-provider-google_v1.20.0_x4: {
plugin.terraform-provider-google_v1.20.0_x4:  "backendType": "SECOND_GEN",
plugin.terraform-provider-google_v1.20.0_x4:  "connectionName": "PROJECT:europe-west2:NAME",
plugin.terraform-provider-google_v1.20.0_x4:  "databaseVersion": "MYSQL_5_7",
plugin.terraform-provider-google_v1.20.0_x4:  "instanceType": "CLOUD_SQL_INSTANCE",
plugin.terraform-provider-google_v1.20.0_x4:  "settings": {
<REMOVED A FEW LINES>
plugin.terraform-provider-google_v1.20.0_x4:   "locationPreference": {},
plugin.terraform-provider-google_v1.20.0_x4:   "maintenanceWindow": {},
plugin.terraform-provider-google_v1.20.0_x4:   "pricingPlan": "PER_USE",
plugin.terraform-provider-google_v1.20.0_x4:   "replicationType": "SYNCHRONOUS",
plugin.terraform-provider-google_v1.20.0_x4:   "settingsVersion": "1",
plugin.terraform-provider-google_v1.20.0_x4:   "tier": "db-n1-standard-1"


plugin.terraform-provider-google_v1.20.0_x4: ---[ RESPONSE ]--------------------------------------
plugin.terraform-provider-google_v1.20.0_x4: HTTP/2.0 400 Bad Request
plugin.terraform-provider-google_v1.20.0_x4: Alt-Svc: quic=":443"; ma=2592000; v="44,43,39,35"
plugin.terraform-provider-google_v1.20.0_x4: Cache-Control: private, max-age=0
plugin.terraform-provider-google_v1.20.0_x4: Content-Type: application/json; charset=UTF-8
plugin.terraform-provider-google_v1.20.0_x4: Date: Tue, 18 Dec 2018 17:44:23 GMT
plugin.terraform-provider-google_v1.20.0_x4: Expires: Tue, 18 Dec 2018 17:44:23 GMT
plugin.terraform-provider-google_v1.20.0_x4: Server: GSE
plugin.terraform-provider-google_v1.20.0_x4: Vary: Origin
plugin.terraform-provider-google_v1.20.0_x4: Vary: X-Origin
plugin.terraform-provider-google_v1.20.0_x4: X-Content-Type-Options: nosniff
plugin.terraform-provider-google_v1.20.0_x4: X-Frame-Options: SAMEORIGIN
plugin.terraform-provider-google_v1.20.0_x4: X-Xss-Protection: 1; mode=block
plugin.terraform-provider-google_v1.20.0_x4:
plugin.terraform-provider-google_v1.20.0_x4: {
plugin.terraform-provider-google_v1.20.0_x4:  "error": {
plugin.terraform-provider-google_v1.20.0_x4:   "errors": [
plugin.terraform-provider-google_v1.20.0_x4:    {
plugin.terraform-provider-google_v1.20.0_x4:     "domain": "global",
plugin.terraform-provider-google_v1.20.0_x4:     "reason": "invalidOperation",
plugin.terraform-provider-google_v1.20.0_x4:     "message": "This operation isn\"t valid for this instance."
plugin.terraform-provider-google_v1.20.0_x4:    }
plugin.terraform-provider-google_v1.20.0_x4:   ],
plugin.terraform-provider-google_v1.20.0_x4:   "code": 400,
plugin.terraform-provider-google_v1.20.0_x4:   "message": "This operation isn\"t valid for this instance."
plugin.terraform-provider-google_v1.20.0_x4:  }
plugin.terraform-provider-google_v1.20.0_x4: }
```

I initially thought this was related to https://github.com/terraform-providers/terraform-provider-google/pull/2630 but it isn't.

**Comment by davefp** [comment]
I'm having the same problem with my database instance. Specifying an explicit maintenance window then running the apply did not fix the issue 😢 

In my case it's trying to do the following update:

```
  settings.0.maintenance_window.0.day:          "0" => "7"
  settings.0.maintenance_window.0.hour:         "0" => "23"
  settings.0.maintenance_window.0.update_track: "" => "stable"
```

**Comment by chrisst** [comment]
@gburiola I'm unable to repro right now with 1.20 or the latest in master. Also I see that there have been a few fixes to sql-instance merged since 1.20 but I'm not sure if this has been fixed without being able to repro on my side.

Can you try to repro against the latest master build of the plugin and see if this is still an issue? If it is, can you add an example config that shows this?

**Comment by gburiola** [comment]
Hey @chrisst.

I just noticed a mistake on our side. Our TF code is autogenerated and uses json instead of HCL.

We were declaring an empty `"maintenance_window":{ },`

Removing the empty `maintenance_window` block fixed the problem.

Apologies for the confusion and thanks for looking into this.
